torch/csrc/jit/passes/onnx/function_extraction.cpp

```diff
@@ -0,0 +1,1101 @@
+    func_n->output(i)->setType(scope_ctx.outputs_[i]->type());
+    scope_ctx.outputs_[i]->replaceAllUsesWith(func_n->output(i));
+  }
+
+  // set constants and attributes of different values as function attributes.
+
+  for (const auto& it : func_ctx.constant_map_) {
+    TORCH_INTERNAL_ASSERT(it.first->kind() == c10::onnx::Constant);
+    TORCH_INTERNAL_ASSERT(it.first->outputs().size() == 1);
+    auto attr_name = func_ctx.FindAttrName(it.first).value();
+    auto val = it.first->t(attr::value);
+    for (auto* n : scope_ctx.nlist_) {
+      if (it.second.find(n) != it.second.end()) {
+        val = n->t(attr::value);
+        break;
+      }
+    }
+    func_n->t_(Symbol::attr(attr_name), val);
+  }
+
+  auto copy_attr =
+      [](Node* a, Node* b, Symbol attr, const std::string& new_name) {
+#define COPY_ATTR(kind)                                \
+  case AttributeKind::kind: {                          \
+    b->kind##_(Symbol::attr(new_name), a->kind(attr)); \
+    break;                                             \
+  }
+        switch (a->kindOf(attr)) {
+          COPY_ATTR(f)
+          COPY_ATTR(fs)
+          COPY_ATTR(i)
+          COPY_ATTR(is)
+          COPY_ATTR(s)
+          COPY_ATTR(ss)
+          COPY_ATTR(t)
+          COPY_ATTR(ts)
+#undef COPY_ATTR
+          case AttributeKind::ival:
+          case AttributeKind::g:
+          case AttributeKind::gs:
+          case AttributeKind::ty:
+          case AttributeKind::tys:
+          case AttributeKind::c:
+          default:
+            TORCH_INTERNAL_ASSERT(
+                false,
+                "Unexpected attribute type ",
+                static_cast<int>(a->kindOf(attr)),
+                " from node ",
+                *a);
+            break;
+        }
+      };
+
+  for (const auto& it : func_ctx.attribute_map_) {
+    auto* ref_n = it.first;
+    for (const auto attr_it : it.second) {
+      const auto& attr = attr_it.first;
+      auto attr_name = func_ctx.FindAttrName(ref_n, attr).value();
+      copy_attr(ref_n, func_n, attr, attr_name);
+      for (auto* n : scope_ctx.nlist_) {
+        if (attr_it.second.find(n) != attr_it.second.end()) {
+          copy_attr(n, func_n, attr, attr_name);
+          break;
+        }
+      }
+    }
+  }
+
+  func_n->insertAfter(last_n);
+
+  return func_n;
+}
+
+void FunctionExtractor::ConvertScopeToFunction(
+    const ScopePtr& scope_key,
+    const scope_list& scope_list,
+    scope_ctx_map& scope_ctxs,
+    const std::shared_ptr<Graph>& graph) {
+  // This function needs to be called always on inner most scopes.
+  // 1. Generate function context, this identifies different constants and
+  // attributes.
+  // 2. Create function definition node, and insert to main graph.
+  // 3. Create function node for each call, and replace subgraph nodes in parent
+  // functions.
+
+  func_ctxs_.insert(std::make_pair(
+      scope_key, new FunctionContext(scope_key, scope_list, scope_ctxs)));
```

Comment:
where to delete? smart pointer?

Reply:
deleted at the last few lines in run(). Used raw pointer to save the additional overhead, since the lifecycle of this object is fairly simple and won't be shared out of this context.